File: src/brs/web/api/http/handler/GetTransaction.java
package brs.web.api.http.handler;

import brs.Blockchain;
import brs.Transaction;
import brs.TransactionProcessor;
import brs.util.Convert;
import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.JSONData;
import brs.web.api.http.common.LegacyDocTag;
import com.google.gson.JsonElement;

import javax.servlet.http.HttpServletRequest;

import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.FULL_HASH_PARAMETER;
import static brs.web.api.http.common.Parameters.TRANSACTION_PARAMETER;

public final class GetTransaction extends ApiServlet.JsonRequestHandler {

  private final TransactionProcessor transactionProcessor;
  private final Blockchain blockchain;

  public GetTransaction(TransactionProcessor transactionProcessor, Blockchain blockchain) {
    super(new LegacyDocTag[] {LegacyDocTag.TRANSACTIONS}, TRANSACTION_PARAMETER, FULL_HASH_PARAMETER);
    this.transactionProcessor = transactionProcessor;
    this.blockchain = blockchain;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) {

    String transactionIdString = Convert.emptyToNull(req.getParameter(TRANSACTION_PARAMETER));
    String transactionFullHash = Convert.emptyToNull(req.getParameter(FULL_HASH_PARAMETER));
    if (transactionIdString == null && transactionFullHash == null) {
      return MISSING_TRANSACTION;
    }

    long transactionId = 0;
    Transaction transaction;
    try {
      if (transactionIdString != null) {
        transactionId = Convert.parseUnsignedLong(transactionIdString);
        transaction = blockchain.getTransaction(transactionId);
      } else {
        transaction = blockchain.getTransactionByFullHash(transactionFullHash);
        if (transaction == null) {
          return UNKNOWN_TRANSACTION;
        }
      }
    } catch (RuntimeException e) {
      return INCORRECT_TRANSACTION;
    }

    if (transaction == null) {
      transaction = transactionProcessor.getUnconfirmedTransaction(transactionId);
      if (transaction == null) {
        return UNKNOWN_TRANSACTION;
      }
      return JSONData.unconfirmedTransaction(transaction);
    } else {
      return JSONData.transaction(transaction, blockchain.getHeight());
    }

  }

}

File: src/main/java/brs/http/GetTransaction.kt
package brs.http

import brs.Blockchain
import brs.Transaction
import brs.TransactionProcessor
import brs.util.Convert
import com.google.gson.JsonElement

import javax.servlet.http.HttpServletRequest

import brs.http.JSONResponses.INCORRECT_TRANSACTION
import brs.http.JSONResponses.MISSING_TRANSACTION
import brs.http.JSONResponses.UNKNOWN_TRANSACTION
import brs.http.common.Parameters.FULL_HASH_PARAMETER
import brs.http.common.Parameters.TRANSACTION_PARAMETER

internal class GetTransaction(private val transactionProcessor: TransactionProcessor, private val blockchain: Blockchain) : APIServlet.JsonRequestHandler(arrayOf(APITag.TRANSACTIONS), TRANSACTION_PARAMETER, FULL_HASH_PARAMETER) {

    internal override fun processRequest(req: HttpServletRequest): JsonElement {

        val transactionIdString = Convert.emptyToNull(req.getParameter(TRANSACTION_PARAMETER))
        val transactionFullHash = Convert.emptyToNull(req.getParameter(FULL_HASH_PARAMETER))
        if (transactionIdString == null && transactionFullHash == null) {
            return MISSING_TRANSACTION
        }

        var transactionId: Long = 0
        var transaction: Transaction?
        try {
            if (transactionIdString != null) {
                transactionId = Convert.parseUnsignedLong(transactionIdString)
                transaction = blockchain.getTransaction(transactionId)
            } else {
                transaction = blockchain.getTransactionByFullHash(transactionFullHash!!)
                if (transaction == null) {
                    return UNKNOWN_TRANSACTION
                }
            }
        } catch (e: RuntimeException) {
            return INCORRECT_TRANSACTION
        }

        if (transaction == null) {
            transaction = transactionProcessor.getUnconfirmedTransaction(transactionId)
            return if (transaction == null) {
                UNKNOWN_TRANSACTION
            } else JSONData.unconfirmedTransaction(transaction)
        } else {
            return JSONData.transaction(transaction, blockchain.height)
        }

    }

}
